[FIX] 예약 내역 목록 조회 N+1 문제 해결

jye326 opened the pull request:
## Issue Number
closed #701 

## As-Is
<!-- 문제 상황 정의 -->
예약 내역 목록 조회 로직의 N+1 문제 발생

## To-Be
<!-- 변경 사항 -->
- 회원 예약 목록 조회 시 기존 `Reservation` 엔터티 반환 대신 `ParticipatedReservationView` DTO를 반환하도록 변경
- JPQL 쿼리로 예약 관련 데이터 조회 최적화
- `ReservationService`의 기존 접근 방식 리팩토링을 통해 불필요한 메서드 제거
- `ParticipatedReservationView` 인터페이스 추가하여 필요한 데이터만 추출 및 반환

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description

jye326 committed:
feat: 회원 예약 조회 최적화 및 DTO 도입

- 회원 예약 목록 조회 시 기존 `Reservation` 엔터티 반환 대신 `ParticipatedReservationView` DTO를 반환하도록 변경
- JPQL 쿼리로 예약 관련 데이터 조회 최적화
- `ReservationService`의 기존 접근 방식 리팩토링을 통해 불필요한 메서드 제거
- `ParticipatedReservationView` 인터페이스 추가하여 필요한 데이터만 추출 및 반환

jye326 committed:
refactor: 테스트 데이터 정렬 방식 수정

- `ReservationServiceTest`의 테스트 데이터 순서 재조정
- `ParticipatedReservationResponse` 리스트 내 항목 위치 변경
- 테스트 가독성 개선 및 일관성 유지

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/ReservationService.java:
```diff
@@ -112,9 +110,17 @@ public PhoneNumberResponse getPhone(Long reservationId) {
-        List<Reservation> memberReservations = reservationRepository.findAllByMenteeId(memberId);
-        return memberReservations.stream()
-                .map(this::generateParticipatedReservationResponse)
+        List<ParticipatedReservationView> views = reservationRepository.findMemberReservationsView(memberId);
+        return views.stream()
+                .map(v -> new ParticipatedReservationResponse(
+                        v.getReservationId(),
+                        v.getMentoringId(),
+                        v.getMentorName(),
+                        v.getMentorProfileImage(),
+                        v.getReservedAt(),
+                        v.getContent(),
+                        v.getStatus(),
+                        v.getIsReviewed()))
                 .toList();
```
p3: 기존 코드 컨벤션대로 ParticipatedReservationView DTO가 컨트롤러에서 쓰인다면 Dto 접미사를 붙이는 것으로 수정되면 좋겠습니다.

ppparkta replied:
p3: 인터페이스 프로젝션 사용으로 인해 서비스 계층의 DTO와 컨트롤러 계층의 DTO가 분리된 점 좋네요! 👍

jye326 replied:
반영했습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/ReservationService.java:
```diff
@@ -142,29 +148,6 @@ private void checkAdminAuthority(Long memberId) {
         }
     }
 
-    private ParticipatedReservationResponse generateParticipatedReservationResponse(Reservation reservation) {
```
p3: 복잡한 로직이 사라져서 깔끔해졌네요!!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/dto/ParticipatedReservationView.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.time.LocalDate;
+
+public interface ParticipatedReservationView {
```
p2: 인터페이스 프로젝션 방식을 처음 봐서 신기하네요. 메서드마다 개행 추가해주시고 네이밍 컨벤션 같이 이야기해보면 좋겠습니다.

ppparkta replied:
p3: 인터페이스 방식을 사용하신 이유 코멘트로 정리해주시면 좋을 것 같아요!

jye326 replied:
인터페이스 기반 프로젝션은 JPA가 런타임에 프록시 객체를 만들고, 내부 getter를 확인하여 쿼리 결과값을 매핑하는 기술입니다. getter의 네이밍 기반 매핑이기 때문에 쿼리 내부의 select 순서 등의 영향을 받지 않을 수 있는 장점이 있습니다.
또한 해당 dto 인터페이스는 직접 구현체를 만들기 위한 목적이 아니고, 내부 getter들은 메서드라기보다는 사실상 필드에 가까운 역할을 하므로 개행을 하지 않았습니다!

ppparkta reviewed (commented):
쿼리DSL로 어떻게 변경될지 기대되네요!! 고생하셨습니다. 코멘트 몇 개만 확인 부탁드립니다.

jye326 committed:
Merge branch 'develop' into fix/701-participated-reservation-n+1

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/mentoring/business/repository/ReservationRepository.java
#	backend/fittoring/src/main/java/fittoring/mentoring/business/service/ReservationService.java

jye326 committed:
refactor: 멘토 예약 조회 로직 최적화

- `ReservationService` 내 멘토 예약 조회 메서드 로직 간소화
- 불필요한 메서드 제거 및 JPQL 쿼리 활용
- 반복문 제거로 성능 및 가독성 개선

dye0p reviewed (approved):
수고하셨어요!

jye326 committed:
refactor: ParticipatedReservationView를 ParticipatedReservationDto로 리팩토링

- `ParticipatedReservationView` 인터페이스를 `ParticipatedReservationDto`로 이름 변경
- 관련 메서드 및 변수명 수정 (ex: `findMemberReservationsView` → `findMemberReservationDtos`)
- `ReservationService` 내 사용된 코드에서 DTO 변경사항 반영
- 가독성과 명확성을 개선하기 위한 리팩토링 적용

ppparkta reviewed (approved):
코멘트 잘 읽었습니다. 마치 builder를 사용하는 것처럼 편리하겠네요. 성능 상 차이가 없으면 기존의 DTO 프로젝션 방식을 인터페이스 기반 프로젝션으로 수정하는 것도 좋을 것 같아요. 수고하셨습니다!
